README.rst
About this module
-----------------
Wrapper module which provides iterable object with infix style method chaining.  

Unreadble::

    >>> filter(lambda y: y % 2, map(lambda x: x + 1, iterable))

Readable::

    >>> iterable = react(iterable)
    >>> iterable.map(lambda x: x + 1).filter(lambda y: y % 2)

How to install
--------------
Requires python2.6 or later (include python3.x).
You need *pip* or *distribute* or *setuptools*::

    $ pip install chainreaction

or use easy_install::

    $ easy_install chainreaction
    
Recent changed
--------------
0.2.0
^^^^^
* Supported python2.6 .
* Changed to use itertools in 'Chainable.filter/map' methods in python2.x .

0.1.2
^^^^^
* Supported 'for' statement.

How to use
----------
The simplest and an only way to use this module is to call the **react** function.  
**react** is a factory function which looks up *'__iter__'* / *'__getitem__'* 
from an given iterable, and returns wrapped object.

Example to use::

    >>> from chainreaction.reactor import react
    >>> react("hello")
    >>> react([1, 2, 3])
    >>> react(iter([1, 2, 3])
    >>> class DictLike(dict):pass
    >>> react(DictLike())

Chainable object support any methods of its wrapped object,
and try to wrap their return value.  
By convention, side effective methods return nothing, but chainable object
returns its side affected object wrapped with appropriate wrappper.  
If iterator was passed into **react**, it would be consumed when it called
any iterative methods.

Example to use::

    >>> tobewrapped = "hello"
    >>> wrapped = react(tobewrapped)
    >>> wrapped = wrapped.upper().map(lambda c: ord(c))
    >>> [chr(c) for c in wrapped.unwrap] = [c for c in tobewrapped.upper()]
    True
    >>> tobewrapped = [1, 2, 3]
    >>> wrapped = react(tobewrapped)
    >>> wrapped.append(4).count()
    4
    >>> len(tobewrapped)
    4

If an iterator was given, this iterator would be consumed
when it called any method::

    >>> tobewrapped = iter([1,2,3])
    >>> wrapped = react(tobewrapped)
    >>> wrapped.count()
    4
    >>> len(list(tobewrapped))
    0
    
Tiny percent encoding implementation example::

    >>> test = bytes("hello world-._~", encoding="UTF-8")
    >>> wrapped = react(test)
    >>> safe = set()
    >>> react("0123456789").foreach(lambda c: safe.add(ord(c)))
    >>> react(range(ord('a'), ord('z') + 1)).foreach(lambda b: safe.add(b))
    >>> react(range(ord('A'), ord('Z') + 1)).foreach(lambda b: safe.add(b))
    >>> react("-._~").foreach(lambda c: safe.add(ord(c)))
    >>> test = wrapped.map(lambda b: b if b > 0 else 256 + b)
    >>> test = test.map(lambda i: '+' if chr(i).isspace() else chr(i) if i in safe else "%{0:x}".format(i))
    >>> test.tostring == "hello+world-._~"
    True
    
Basic API
---------
* unwrap
    Retrieves iterable from wrapped object.
* dump
    Elements dump.
* isiterator
    True if wrapped object is iterator.
* tostring
    Returns almost same with "".join((str(x) for x in iterable)). This method always returns str object. str wrapped returns unwrapped, bytes or bytearray wrapped returns str converted, dict wrapped returns "key=value" pairs joined with "&".
* foreach(f)
    Applies a given f to all elements.
* filter(pred)
    Selects all elements which satisfy a given predicate.
* map(f)
    Builds a new collection by applying a given f.
* forall(pred)
    True if all elements satisfy a given predicate.
* forany(pred)
    True if any elements satisfy a given predicate.
* dropwhile(pred)
    Drops longest prefix of elements that satisfy a given predicate.
* dropright(pred)
    Drops longest suffix of elements that satisfy a given predicate.
* takwhile(pred)
    Takes longest prefix of elements that satisfy a given predicate.
* takeright(pred)
    Takes longest suffix of elements that satisfy a given predicate.
* mkstring(joiner="")
    Returns wrapped str object using a joiner string.
* counts(pred=lambda x: True)
    Counts the number of elements that satisfy a given predicate.
* contains(key)
    Tests whether this wrapped object contains a given key as an element.
* reduce(f)
    Returns a value(not wrapped) using a given f.
    
iterator specific API
^^^^^^^^^^^^^^^^^^^^^
* tolist
    Returns a new list wrapped.
* totuple
    Returns a new tuple wrapped.
* toset
    Returns a new set wrapped.
    
str, bytes, bytearray specific API
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
* reverse
    Returns a new reversed str wrapped.

seq specific API
^^^^^^^^^^^^^^^^
* accumulate(f)
    Returns a seq of accumulated value.
* reverse
    Returns a new reversed seq wrapped.
* sort
    Returns a new sorted seq wrapped.
* toset
    Returns a new set wrapped.

set specific API
^^^^^^^^^^^^^^^^
* min
    Returns a minimum value.
* max
    Returns a maximum value.

HISTORY.rst
History
-------
0.2.0
^^^^^
* Supported python2.6!
* Changed to use itertools in 'Chainable.filter/map' methods in python2.

0.1.2
^^^^^
* Supported 'for' statement.

test/utils.py
""" Quoted from activestate.
http://code.activestate.com/recipes/576693/
"""

## {{{ http://code.activestate.com/recipes/576693/ (r9)
# Backport of OrderedDict() class that runs on Python 2.4, 2.5, 2.6, 2.7 and pypy.
# Passes Python2.7's test suite and incorporates all the latest updates.

try:
    from thread import get_ident as _get_ident
except ImportError:
    from dummy_thread import get_ident as _get_ident

try:
    from _abcoll import KeysView, ValuesView, ItemsView
except ImportError:
    pass


class OrderedDict(dict):
    'Dictionary that remembers insertion order'
    # An inherited dict maps keys to values.
    # The inherited dict provides __getitem__, __len__, __contains__, and get.
    # The remaining methods are order-aware.
    # Big-O running times for all methods are the same as for regular dictionaries.

    # The internal self.__map dictionary maps keys to links in a doubly linked list.
    # The circular doubly linked list starts and ends with a sentinel element.
    # The sentinel element never gets deleted (this simplifies the algorithm).
    # Each link is stored as a list of length three:  [PREV, NEXT, KEY].

    def __init__(self, *args, **kwds):
        '''Initialize an ordered dictionary.  Signature is the same as for
        regular dictionaries, but keyword arguments are not recommended
        because their insertion order is arbitrary.

        '''
        if len(args) > 1:
            raise TypeError('expected at most 1 arguments, got %d' % len(args))
        try:
            self.__root
        except AttributeError:
            self.__root = root = []                     # sentinel node
            root[:] = [root, root, None]
            self.__map = {}
        self.__update(*args, **kwds)

    def __setitem__(self, key, value, dict_setitem=dict.__setitem__):
        'od.__setitem__(i, y) <==> od[i]=y'
        # Setting a new item creates a new link which goes at the end of the linked
        # list, and the inherited dictionary is updated with the new key/value pair.
        if key not in self:
            root = self.__root
            last = root[0]
            last[1] = root[0] = self.__map[key] = [last, root, key]
        dict_setitem(self, key, value)

    def __delitem__(self, key, dict_delitem=dict.__delitem__):
        'od.__delitem__(y) <==> del od[y]'
        # Deleting an existing item uses self.__map to find the link which is
        # then removed by updating the links in the predecessor and successor nodes.
        dict_delitem(self, key)
        link_prev, link_next, key = self.__map.pop(key)
        link_prev[1] = link_next
        link_next[0] = link_prev

    def __iter__(self):
        'od.__iter__() <==> iter(od)'
        root = self.__root
        curr = root[1]
        while curr is not root:
            yield curr[2]
            curr = curr[1]

    def __reversed__(self):
        'od.__reversed__() <==> reversed(od)'
        root = self.__root
        curr = root[0]
        while curr is not root:
            yield curr[2]
            curr = curr[0]

    def clear(self):
        'od.clear() -> None.  Remove all items from od.'
        try:
            for node in self.__map.itervalues():
                del node[:]
            root = self.__root
            root[:] = [root, root, None]
            self.__map.clear()
        except AttributeError:
            pass
        dict.clear(self)

    def popitem(self, last=True):
        '''od.popitem() -> (k, v), return and remove a (key, value) pair.
        Pairs are returned in LIFO order if last is true or FIFO order if false.

        '''
        if not self:
            raise KeyError('dictionary is empty')
        root = self.__root
        if last:
            link = root[0]
            link_prev = link[0]
            link_prev[1] = root
            root[0] = link_prev
        else:
            link = root[1]
            link_next = link[1]
            root[1] = link_next
            link_next[0] = root
        key = link[2]
        del self.__map[key]
        value = dict.pop(self, key)
        return key, value

    # -- the following methods do not depend on the internal structure --

    def keys(self):
        'od.keys() -> list of keys in od'
        return list(self)

    def values(self):
        'od.values() -> list of values in od'
        return [self[key] for key in self]

    def items(self):
        'od.items() -> list of (key, value) pairs in od'
        return [(key, self[key]) for key in self]

    def iterkeys(self):
        'od.iterkeys() -> an iterator over the keys in od'
        return iter(self)

    def itervalues(self):
        'od.itervalues -> an iterator over the values in od'
        for k in self:
            yield self[k]

    def iteritems(self):
        'od.iteritems -> an iterator over the (key, value) items in od'
        for k in self:
            yield (k, self[k])

    def update(*args, **kwds):
        '''od.update(E, **F) -> None.  Update od from dict/iterable E and F.

        If E is a dict instance, does:           for k in E: od[k] = E[k]
        If E has a .keys() method, does:         for k in E.keys(): od[k] = E[k]
        Or if E is an iterable of items, does:   for k, v in E: od[k] = v
        In either case, this is followed by:     for k, v in F.items(): od[k] = v

        '''
        if len(args) > 2:
            raise TypeError('update() takes at most 2 positional '
                            'arguments (%d given)' % (len(args),))
        elif not args:
            raise TypeError('update() takes at least 1 argument (0 given)')
        self = args[0]
        # Make progressively weaker assumptions about "other"
        other = ()
        if len(args) == 2:
            other = args[1]
        if isinstance(other, dict):
            for key in other:
                self[key] = other[key]
        elif hasattr(other, 'keys'):
            for key in other.keys():
                self[key] = other[key]
        else:
            for key, value in other:
                self[key] = value
        for key, value in kwds.items():
            self[key] = value

    __update = update  # let subclasses override update without breaking __init__

    __marker = object()

    def pop(self, key, default=__marker):
        '''od.pop(k[,d]) -> v, remove specified key and return the corresponding value.
        If key is not found, d is returned if given, otherwise KeyError is raised.

        '''
        if key in self:
            result = self[key]
            del self[key]
            return result
        if default is self.__marker:
            raise KeyError(key)
        return default

    def setdefault(self, key, default=None):
        'od.setdefault(k[,d]) -> od.get(k,d), also set od[k]=d if k not in od'
        if key in self:
            return self[key]
        self[key] = default
        return default

    def __repr__(self, _repr_running={}):
        'od.__repr__() <==> repr(od)'
        call_key = id(self), _get_ident()
        if call_key in _repr_running:
            return '...'
        _repr_running[call_key] = 1
        try:
            if not self:
                return '%s()' % (self.__class__.__name__,)
            return '%s(%r)' % (self.__class__.__name__, self.items())
        finally:
            del _repr_running[call_key]

    def __reduce__(self):
        'Return state information for pickling'
        items = [[k, self[k]] for k in self]
        inst_dict = vars(self).copy()
        for k in vars(OrderedDict()):
            inst_dict.pop(k, None)
        if inst_dict:
            return (self.__class__, (items,), inst_dict)
        return self.__class__, (items,)

    def copy(self):
        'od.copy() -> a shallow copy of od'
        return self.__class__(self)

    @classmethod
    def fromkeys(cls, iterable, value=None):
        '''OD.fromkeys(S[, v]) -> New ordered dictionary with keys from S
        and values equal to v (which defaults to None).

        '''
        d = cls()
        for key in iterable:
            d[key] = value
        return d

    def __eq__(self, other):
        '''od.__eq__(y) <==> od==y.  Comparison to another OD is order-sensitive
        while comparison to a regular mapping is order-insensitive.

        '''
        if isinstance(other, OrderedDict):
            return len(self) == len(other) and self.items() == other.items()
        return dict.__eq__(self, other)

    def __ne__(self, other):
        return not self == other

    # -- the following methods are only used in Python 2.7 --

    def viewkeys(self):
        "od.viewkeys() -> a set-like object providing a view on od's keys"
        return KeysView(self)

    def viewvalues(self):
        "od.viewvalues() -> an object providing a view on od's values"
        return ValuesView(self)

    def viewitems(self):
        "od.viewitems() -> a set-like object providing a view on od's items"
        return ItemsView(self)


test/test_ref.py
'''
Created on 2011/07/11

@author: Keita Oouchi
'''
import sys, os

sys.path.append("./src")

import unittest
from chainreaction.reactor import react, Chainable
from sys import getrefcount as ref

class TestReactor(unittest.TestCase):
    def setUp(self):
        self.ispy3 = True if sys.version >= '3.0' else False

    def iterref(self, obj):
        #alignment
        startref = ref(obj)
        wrap = react(obj)
        self.assertEqual(ref(obj), startref + 1)

        #iter(wrap) => wrap.__iter__ => iter(wrap._iterable)
        itered = iter(wrap)
        self.assertEqual(ref(obj), startref + 2)
        itered = None
        self.assertEqual(ref(obj), startref + 1)

        #'for' statement use __iter__, __iter__ calls iter(self._iterable).
        for temp in wrap:
            self.assertEqual(ref(obj), startref + 2)
        self.assertEqual(ref(obj), startref + 1)
        self.assertEqual(ref(wrap), 2)

    def dictiterref(self, obj):
        self.assertTrue(issubclass(obj.__class__, dict))
        #alignment
        startref = ref(obj)
        wrap = react(obj)
        self.assertEqual(ref(obj), startref + 1)
        """
        In python3.x, 'obj.items()' returns 'dict_items' object.
        By contrast, python2.x returns 'list' object.
        'dict_items' object is a 'view' object, and this is nearly iterator,
        This iterator holds a reference to obj, so reference count is incremented.
        """
        itered = iter(wrap)
        self.assertFalse(issubclass(itered.__class__, Chainable))
        expect = 2 if self.ispy3 else 1
        self.assertEqual(ref(obj), startref + expect)
        itered = None
        self.assertEqual(ref(obj), startref + 1)

        #'for' statement use __iter__, __iter__ calls iter(self._iterator).
        for temp in wrap:
            self.assertEqual(ref(obj), startref + expect)
        self.assertEqual(ref(obj), startref + 1)
        self.assertEqual(ref(wrap), 2)

    def testStr(self):
        teststr = "reftest"
        self.iterref(teststr)
        startref = ref(teststr)
        wrap = react(teststr)
        self.assertEqual(ref(teststr), startref + 1)
        #filter, takewhile, dropwhile, etc refer to new string.
        refalign = 1
        filtered = wrap.filter(lambda c: ord(c) > 0)
        self.assertEqual(ref(teststr), startref + refalign)
        taken = wrap.takewhile(lambda c: ord(c) > 0)
        self.assertEqual(ref(teststr), startref + refalign)
        dropped = wrap.dropwhile(lambda c: ord(c) < 0)
        self.assertEqual(ref(teststr), startref + refalign)
        tr = wrap.takeright(lambda c: ord(c) > 0)
        self.assertEqual(ref(teststr), startref + refalign)

        #map refer to iterator-like
        mapped = wrap.map(lambda c: chr(ord(c) + 1))
        refalign += 1
        self.assertEqual(ref(teststr), startref + refalign)
        self.assertEqual(ref(wrap), 2)
        self.assertEqual(ref(filtered.unwrap), 2)
        self.assertEqual(ref(mapped.unwrap), 2)
        self.assertEqual(ref(taken.unwrap), 2)

        upped = wrap.upper()
        self.assertEqual(ref(teststr), startref + refalign)
        self.assertEqual(ref(wrap), 2)

        wrap = wrap.reverse
        refalign -= 1
        self.assertEqual(ref(teststr), startref + refalign)
        self.assertEqual(ref(wrap.unwrap), 2)

        mapped = None
        self.assertEqual(ref(teststr), startref)

    def testSeq(self):
        testseq = [1, 2, 3]
        self.iterref(testseq)
        startref = ref(testseq)
        wrap = react(testseq)
        self.assertEqual(ref(testseq), startref + 1)
        #filter, map, takewhile, etc... refer to new seq.
        filted = wrap.filter(lambda i: i > 0)
        self.assertEqual(filted.unwrap.__class__, list)
        mapped = wrap.map(lambda i: i + 1)
        taken = wrap.takewhile(lambda i: i > 0)
        self.assertEqual(ref(testseq), startref + 1)
        self.assertEqual(ref(wrap), 2)
        self.assertEqual(ref(filted.unwrap), 2) #passed to ref and filted internal.
        self.assertEqual(ref(mapped.unwrap), 2)
        self.assertEqual(ref(taken.unwrap), 2)

        appended = wrap.append(4)
        self.assertEqual(ref(testseq), startref + 2) #wrap, appended

        sortedwrap = wrap.sort
        self.assertEqual(ref(testseq), startref + 2) #wrap, appended
        self.assertEqual(ref(wrap), 2)

        wrap = wrap.reverse
        self.assertEqual(ref(testseq), startref + 1) #appended
        self.assertEqual(ref(wrap.unwrap), 2)

        appended = None
        self.assertEqual(ref(testseq), startref)

    def testSet(self):
        testset = set([1, 2, 3])
        self.iterref(testset)
        startref = ref(testset)
        wrap = react(testset)
        self.assertEqual(ref(testset), startref + 1)
        #filter, map, takewhile, etc... refer to new set.
        filted = wrap.filter(lambda i: i > 0)
        mapped = wrap.map(lambda i: i + 1)
        taken = wrap.takewhile(lambda i: i > 0)
        self.assertEqual(ref(testset), startref + 1)
        self.assertEqual(ref(wrap), 2)
        self.assertEqual(ref(filted.unwrap), 2) #passed to ref and filted internal.
        self.assertEqual(ref(mapped.unwrap), 2)
        self.assertEqual(ref(taken.unwrap), 2)

        unionized = wrap.union(set([4, 5, 6]))
        self.assertEqual(ref(testset), startref + 1)
        self.assertEqual(ref(wrap), 2)

        wrap = wrap.union(set([4, 5, 6]))
        self.assertEqual(ref(testset), startref)
        self.assertEqual(ref(wrap.unwrap), 2)

    def testDict(self):
        testdict = {"k1":"v1", "k2":""}
        self.dictiterref(testdict)
        startref = ref(testdict)
        wrap = react(testdict)
        self.assertEqual(ref(testdict), startref + 1)
        #filter, map, takewhile, etc... refer to new dict.
        filted = wrap.filter(lambda t: t[1])
        mapped = wrap.map(lambda t: (t[0], t[0] + "=" + t[1]))
        taken = wrap.takewhile(lambda t: t[1])
        self.assertEqual(ref(testdict), startref + 1)
        self.assertEqual(ref(wrap), 2)
        self.assertEqual(ref(filted.unwrap), 2) #passed to ref and filted internal.
        self.assertEqual(ref(mapped.unwrap), 2)
        self.assertEqual(ref(taken.unwrap), 2)

        updated = wrap.update(("k3", "v3"))
        self.assertEqual(ref(testdict), startref + 2) #wrap, updated
        self.assertEqual(ref(wrap), 2)
        self.assertEqual(updated.unwrap, testdict)

        wrap = wrap.update(("k3", "v3"))
        self.assertEqual(ref(testdict), startref + 2)
        self.assertEqual(ref(wrap.unwrap), startref + 2)

    def testIter(self):
        testseq = [1, 2, 3]
        testit = iter(testseq)
        self.iterref(testit)
        startref = ref(testit)
        wrap = react(testit)
        self.assertEqual(ref(testit), startref + 1)
        base = startref + 1
        refalign = 0
        #filter, map, takewhile, dropwhile returns IterChain.
        filted = wrap.filter(lambda i: i > 0)
        refalign += 1
        self.assertEqual(ref(testit), base + refalign)
        mapped = wrap.map(lambda i: i + 1)
        refalign += 1
        self.assertEqual(ref(testit), base + refalign)
        taken = wrap.takewhile(lambda i: i > 0)
        refalign += 1
        self.assertEqual(ref(testit), base + refalign)
        droped = wrap.dropwhile(lambda i: i < 0)
        refalign += 1
        self.assertEqual(ref(testit), base + refalign)

        #takeright, dropright returns SeqChain.
        tr = wrap.takeright(lambda i: i > 0)
        dr = wrap.dropright(lambda i: i < 0)
        self.assertEqual(ref(testit), base + refalign)

        taken = None
        refalign -= 1
        self.assertEqual(ref(testit), base + refalign)
        right = wrap.takeright(lambda i: i < 0)
        self.assertEqual(ref(testit), base + refalign)
        self.assertEqual(ref(wrap), 2)

def suite():
  suite = unittest.TestSuite()
  suite.addTests(unittest.makeSuite(TestReactor))
  return suite

if __name__ == "__main__":
    try:
        from sys import getrefcount
        suite()
    except:
        pass


src/chainreaction/reactor.py
"""Wrapper module which provides iterable object with infix style method chain.  

Unreadble::

    >>> filter(lambda y: y % 2, map(lambda x: x + 1, iterable))

Readable::

    >>> iterable = react(iterable)
    >>> iterable.map(lambda x: x + 1).filter(lambda y: y % 2)

How to install
--------------
Requires python2.6 or later (include python3.x).
You need *pip* or *distribute* or *setuptools*::

    $ pip install chainreaction

or use easy_install::

    $ easy_install chainreaction

How to use
----------
The simplest and an only way to use this module is to call the **react** function.  
**react** is a factory function which looks up *'__iter__'* / *'__getitem__'* 
from an given iterable, and returns wrapped object.

Example to use::

    >>> from chainreaction.reactor import react
    >>> react("hello")
    >>> react([1, 2, 3])
    >>> react(iter([1, 2, 3])
    >>> class DictLike(dict):pass
    >>> react(DictLike())

Chainable object support any methods of its wrapped object,
and try to wrap their return value.  
By convention, side effective methods return nothing, but chainable object
returns its side affected object wrapped with appropriate wrappper.  
If iterator was passed into **react**, it would be consumed when it called
any iterative methods.

Example to use::

    >>> tobewrapped = "hello"
    >>> wrapped = react(tobewrapped)
    >>> wrapped = wrapped.upper().map(lambda c: ord(c))
    >>> [chr(c) for c in wrapped.unwrap] = [c for c in tobewrapped.upper()]
    True
    >>> tobewrapped = [1, 2, 3]
    >>> wrapped = react(tobewrapped)
    >>> wrapped.append(4).count()
    4
    >>> len(tobewrapped)
    4

If an iterator was given, this iterator would be consumed
when it called any method::

    >>> tobewrapped = iter([1,2,3])
    >>> wrapped = react(tobewrapped)
    >>> wrapped.count()
    4
    >>> len(list(tobewrapped))
    0
    
Tiny percent encoding implementation example::

    >>> test = bytes("hello world-._~", encoding="UTF-8")
    >>> wrapped = react(test)
    >>> safe = set()
    >>> react("0123456789").foreach(lambda c: safe.add(ord(c)))
    >>> react(range(ord('a'), ord('z') + 1)).foreach(lambda b: safe.add(b))
    >>> react(range(ord('A'), ord('Z') + 1)).foreach(lambda b: safe.add(b))
    >>> react("-._~").foreach(lambda c: safe.add(ord(c)))
    >>> test = wrapped.map(lambda b: b if b > 0 else 256 + b)
    >>> test = test.map(lambda i: '+' if chr(i).isspace() else chr(i) if i in safe else "%{0:x}".format(i))
    >>> test.tostring == "hello+world-._~"
    True

@author: Keita Oouchi
@email : keita.oouchi@gmail.com
"""

import itertools
import functools
import collections

__all__ = ["react"]

import sys
if sys.version < '3.0.0':
    #In py3.x, iterator's "next" method was changed to "__next__".
    _ITER_NEXT = "next"
    _MAP = itertools.imap
    _FILTER = itertools.ifilter
else:
    _ITER_NEXT = "__next__"
    _MAP = map
    _FILTER = filter

def react(iterable):
    """Factory function creating Chainable object.

    If iterable supports iterator or sequencial access protocol,
    wrap it with Chainable class and provide some chaining style method.
    """
    if isinstance(iterable, collections.Iterable):
        if isinstance(iterable, Chainable):
            return iterable
        elif isinstance(iterable, collections.Iterator):
            return IterChain(iterable)
        elif hasattr(iterable, "__getitem__"):
            #isinstance(bytearray(), collections.Sequence) => True if py3.x else False
            if isinstance(iterable, (str, bytes, bytearray)):
                return StrChain(iterable)
            elif isinstance(iterable, collections.Mapping):
                return DictChain(iterable)
            else:
                return SeqChain(iterable)
        elif isinstance(iterable, collections.Set):
            return SetChain(iterable)
        else:
            return Chainable(iterable)
    else:
        template = "Chainable does not support {0}. given argument:{1}:{0}"
        raise AttributeError(template.format(type(iterable), iterable))

"""
def react(iterable):
    if isinstance(iterable, collections.Iterable):
        if isinstance(iterable, Chainable):
            return iterable
        elif isinstance(iterable, collections.Iterator):
            return IterChain(iterable)
        elif issubclass(type(iterable), (str, bytes, bytearray)):
            return StrChain(iterable)
        elif isinstance(iterable, collections.Sequence):
            return SeqChain(iterable)
        elif isinstance(iterable, collections.Mapping):
            return DictChain(iterable)
        elif isinstance(iterable, collections.Set):
            return SetChain(iterable)
        else:
            return Chainable(iterable)
    else:
        template = "Chainable does not support {0}. given argument:{1}:{0}"
        raise AttributeError(template.format(type(iterable), iterable))
"""

def _chainreact(__getattr__):
    """Decorator function used in Chainable's __getattr__ method.
    
    Chainable object support methods of its wrapped objects, and try to keep
    its return value also chainable.
    """
    def containment(*methodname):
        self, methodobj = __getattr__(*methodname)
        def reaction(*realargs):
            result = methodobj(*realargs)
            result = result if result else self
            return react(result)
        return reaction
    return containment


class Chainable(object):
    """ This class is a base class provides chainable interface.
    
    Iterator object would be wrapped with this class.
    """

    def __init__(self, iterable):
        #reference duplication!
        self._iterable = iterable
        self.type = type(iterable)

    def __getattribute__(self, name):
        #Support 'for' statement.
        if name == _ITER_NEXT:
            next = getattr(self._iterator, _ITER_NEXT)
            return next(name)
        else:
            return object.__getattribute__(self, name)

    @_chainreact
    def __getattr__(self, name):
        """Wrap return value using '_chanreact'."""
        return self._iterable, getattr(self._iterable, name)

    def __repr__(self):
        return "'{0}':{1}".format(self.unwrap, type(self).__name__)

    def _constract(self, iterable):
        """This method would be called when other methods, excluding methods of
        wrapped object, was called and wrap their return value if possible.
        """
        try:
            if self.isiterator:
                result = iterable
            else:
                result = self.type(iterable)
        except TypeError:
            result = iterable
        finally:
            return result

    def __enter__(self):
        result = self._iterable.__enter__()
        if result:
            return self

    def __exit__(self, *args):
        result = self._iterable.__exit__(args)
        if result:
            return result

    def __iter__(self):
        return iter(self._iterator)

    @property
    def isiterator(self):
        """Check whether given iterable is iterator object."""
        return isinstance(self._iterable, collections.Iterator)

    @property
    def unwrap(self):
        """Return wrapped object."""
        return self._iterable

    @property
    def dump(self):
        dump_template = "[Dumped Information]\n{type}\n{iterable}\n{tostring}"
        dumped_self = {"type":type(self).__name__,
                       "iterable":self.type.__name__,
                       "tostring":self.tostring}
        return dump_template.format(**dumped_self)

    @property
    def _iterator(self):
        """If wrapped obj was subclass of dict, this would return
        self._iterable.items()."""
        return self._iterable

    @property
    def tostring(self):
        return "".join((str(x) for x in self._iterator))

    def foreach(self, f):
        for x in self._iterator:
            f(x)

    def filter(self, pred):
        result = _FILTER(pred, self._iterator)
        return react(self._constract(result))

    def map(self, f):
        result = _MAP(f, self._iterator)
        return react(self._constract(result))

    def forall(self, pred):
        """True if whole elements satisfy a given predicate."""
        for x in self._iterator:
            if not pred(x):
                return False
        return True

    def forany(self, pred):
        """True if any elements satisfy a given predicate."""
        for x in self._iterator:
            if pred(x):
                return True
        return False

    def dropwhile(self, pred):
        """Drops longest prefix of elements that satisfy a given predicate."""
        result = itertools.dropwhile(pred, self._iterator)
        return react(self._constract(result))

    def dropright(self, pred):
        """Drops longest suffix of elements that satisfy a given predicate."""
        lis = [x for x in self._iterator]
        size = len(lis)
        passed = size
        tf = True
        while tf and passed > 0:
            passed -= 1
            tf = pred(lis[passed])
        result = iter(lis[0:passed + 1])
        return react(self._constract(result))

    def takewhile(self, pred):
        """Takes longest prefix of elements that satisfy a given predicate."""
        result = itertools.takewhile(pred, self._iterator)
        return react(self._constract(result))

    def takeright(self, pred):
        """Takes longest suffix of elements that satisfy a given predicate."""
        lis = [x for x in self._iterator]
        size = len(lis)
        passed = size
        tf = True
        while tf and passed > 0:
            passed -= 1
            tf = pred(lis[passed])
        result = iter(lis[passed + 1:size])
        return react(self._constract(result))

    def mkstring(self, joiner=""):
        """Returns wrapped str object using a joiner string."""
        strlist = (str(x) for x in self._iterator)
        return react(joiner.join(strlist))

    def counts(self, pred=lambda x: True):
        """Counts the number of elements that satisfy a given predicate."""
        result = 0
        for x in self._iterator:
            if pred(x):
                result += 1
        return result

    def contains(self, key):
        """Tests whether this wrapped object contains a given key as an element."""
        return key in self._iterator

    def reduce(self, f):
        """End of chain reaction"""
        return functools.reduce(f, self._iterator)

class IterChain(Chainable):

    def _constract(self, iterable):
        return iterable

    @property
    def tolist(self):
        return react(list(self._iterable))

    @property
    def totuple(self):
        return react(tuple(self._iterable))

    @property
    def toset(self):
        return react(set(self._iterable))

class StrChain(Chainable):

    def __getitem__(self, key):
        return react(self._iterable[key])

    def _constract(self, iterable):
        return "".join((str(x) for x in iterable))

    @property
    def reverse(self):
        result = reversed(self._iterable)
        return react(self._constract(result))

    @property
    def tostring(self):
        if self.type == str:
            return self._iterable
        else:
            temp = self._iterable.decode()
            if type(temp) == str:
                return temp
            else:
                return temp.encode()

    def map(self, f):
        """Returns IterChain or SeqChain(in python2.x)"""
        result = _MAP(f, self._iterable)
        return react(result)

    def mkstring(self, joiner=""):
        return react(joiner.join(self._iterable))

class SeqChain(Chainable):

    def __getitem__(self, key):
        """a[i:j] => a.__getitem__(slice(i,j))"""
        item = self._iterable[key]
        try:
            result = react(item)
        except AttributeError:
            result = item
        finally:
            return result

    @property
    def reverse(self):
        result = reversed(self._iterable)
        return react(self._constract(result))

    @property
    def sort(self):
        result = sorted(self._iterable)
        return react(self._constract(result))

    @property
    def toset(self):
        return react(set(self._iterable))

    def accumulate(self, f):
        """Returns a seq of accumulated value."""
        def inner():
            it = iter(self._iterable)
            total = next(it)
            yield total
            for elem in it:
                total = f(total, elem)
                yield total
        return react(self._constract(inner()))

class SetChain(Chainable):

    @property
    def min(self):
        return min(self._iterable)

    @property
    def max(self):
        return max(self._iterable)

class DictChain(Chainable):

    def __init__(self, iterable):
        """DictChain internally uses list of tuple(key,value) as its iterator.
        [example]
        >>> dictchain.map(lambda t: (t[0], t[1] * 2))
        """
        self._iterable = iterable
        self.type = type(iterable)

    @property
    def _iterator(self):
        return self._iterable.items()

    def __getitem__(self, key):
        item = self._iterable[key]
        try:
            result = react(item)
        except AttributeError:
            result = item
        finally:
            return result

    @property
    def tostring(self):
        key_value_pairs = ("{0}={1}".format(str(t[0]), str(t[1])) for t in self._iterator)
        return "&".join(key_value_pairs)

    def mkstring(self, joiner="&"):
        key_value_pairs = ("{0}={1}".format(str(t[0]), str(t[1])) for t in self._iterator)
        return react(joiner.join(key_value_pairs))
